checkout-core/src/main/java/com/adyen/checkout/core/util/FileDownloader.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2022 Adyen N.V.
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by atef on 16/11/2022.
+ */
+
+package com.adyen.checkout.core.util
+
+import android.Manifest
+import android.content.ContentValues
+import android.content.Context
+import android.os.Build
+import android.provider.MediaStore
+import androidx.annotation.RequiresApi
+import androidx.annotation.RequiresPermission
+import androidx.annotation.RestrictTo
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.io.FileOutputStream
+import java.net.MalformedURLException
+import java.net.URL
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+@Suppress("LongParameterList")
+class FileDownloader {
+
+    @RequiresPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+    suspend fun download(
+        context: Context,
+        stringUrl: String,
+        fileName: String,
+        filePath: String,
+        mimeType: String? = null,
+        notifyState: () -> Unit
+    ) {
+        val url = stringUrl.toULR() ?: return
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            downloadAndSaveApi29AndAbove(context, url, fileName, filePath, mimeType ?: "*/*", notifyState)
+        } else {
+            downloadAndSaveApi28AndBelow(context, url, fileName, filePath, notifyState)
+        }
+    }
+
+    @RequiresPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+    private suspend fun downloadAndSaveApi28AndBelow(
+        context: Context,
+        url: URL,
+        fileName: String,
+        filePath: String,
+        notifyState: () -> Unit
```

Comment:
Instead of callbacks, have you tried using a `suspendCoroutine` which allows transforming a callback into a coroutine? [Here's an example](https://stackoverflow.com/questions/48552925/existing-3-function-callback-to-kotlin-coroutines/48562175#48562175)

Reply:
what's the benefit of using `suspendCoroutine` here?

Reply:
Your method is a suspend function but still uses a callback. One main benefit from coroutines and suspend functions is to [replace callbacks](https://levelup.gitconnected.com/avoid-callback-hell-with-kotlin-coroutines-7a72a2a8a701). With `suspendCoroutine` you can remove your callback and return a value with your method instead.